reference/language-concepts/namespaces.markdown

```diff
@@ -6,104 +6,76 @@ sorting: 100
+declared or until the end of the file.
 
-```cf3
-    body file control
-    {
-       namespace => "myspace";
-    }
-```
+[%CFEngine_include_example(namespace_declaration.cf)%]
 
-### Accessing syntax elements between namespaces and the default namespace
+**Notes:**
 
-To distinguish the bundle `mymethod` in the default namespace from one in
-another namespace, you prefix the bundle name with the namespace, separated by
-a colon.
+- Multiple namespaces can be declared within the same file
+- The same namespace can be declared in multiple files
+- The same namespace can be declared in the same file multiple times
 
-```cf3
-    methods:
+## Methods|usebundle
 
-      "namespace demo" usebundle => myspace:mymethod("arg1");
-      "namespace demo" usebundle => mymethod("arg1","arg2");
-```
+Methods promises assume you are referring to a bundle in the same namespace as
+the promiser. To refer to a bundle in another namespace you *must* specify the
+namespace by prefixing the bundle name with the namespace followed by a colon
+(`:`).
 
-To distinguish a body from one in another namespace, you can prefix the body name with the namespace, separated by a colon.
+[%CFEngine_include_example(namespace_methods-usebundle.cf)%]
 
-```cf3
-    files:
-       "/file"
-          create => "true",
-           perms => name1:settings;
-```
+## Bodies
 
-If you don't make any namespace declarations, you'll be in the
-`default` namespace.  Bundles, bodies, classes, and variables from the
-`default` namespace can be accessed like any other:
+Bodies are assumed to be within the same namespace as the promiser. To use a body from another namespace the namespace must be specified by prefixing the body name with the namespace followed by a colon (`:`).
 
-```cf3
-    files:
-      "/file"
-         create => "true",
-          perms => default:settings;
-```
+A common mistake is forgetting to specify `default:` when using bodies from the standard library which resides in the `default` namespace.
 
-If you use the standard library from your own namespace, remember to
-specify this `default:` prefix.
+[%CFEngine_include_example(namespace_bodies.cf)%]
 
-To access classes, variables, or meta-data in bundles in a different namespace, use the
-colon as a namespace prefix:
+## Variables
 
-    $(namespace:bundle.variable)
-    $(namespace:bundle_meta.variable)
+Variables (except for Special Variables) are assumed to be within the same scope
+as the promiser but can also be referenced fully qualified with the namespace.
 
-**Note** that this means that if you are in a namespace that's not `default`, you *must* qualify classes from `default` fully:
+[%CFEngine_include_example(namespace_variable_references.cf)%]
 
-    default:myclass::
-    "do something" if => "default:myotherclass";
+[Special variables][Special Variables] are always accessible without a namespace
+  prefix. For example, `this`, `mon`, `sys`, and `const` fall in this category.
```

Comment:
actually here it might be good to say something about how these examples are already a prefix but with a dot? How does that work exactly? Seems like they are more like bundles/bodies these special variables?

Reply:
Hummm I don't understand

Reply:
So like `this.bundle` what is the `this` part? Looks like a bundle to me since it is separated by a dot `.` from the variable name `bundle`.

Reply:
I think that I would normally call the `this` part of `this.bundle` a *bundle scope* or a *context* but maybe those names aren't the best. it's not (well, there should not be any bundle named `this`) and *context* is also a term used to refer to a class or a class expression. It's documented in [Special Variables -> this](https://docs.cfengine.com/docs/3.18/reference-special-variables-this.html).

Presence of a `.` alone doesn't tell you what something is. You have to look at the surrounding context.

For example:
`foo.bar` could be part of a variable reference (`$(foo.bar)`) or it could be part of a class expression (`foo.bar::`). But it could as well be fed to something like `if => "foo.bar"` in which case it's a class expression. Or if it was something like `usebundle => my_bundle('foo.bar')` you would have to go look inside `my_bundle` and determine how it's being used. It could be fed to something like `if => "$(my_bundle_param1)"` in which case it's a class expression, or it could be used with something like `"myvar" string => "$($(my_bundle_param1))"` which would be a variable dereference.

Reply:
There is a link there to special variables so that is good. Maybe include one example just so folks get the basic gist without following the link.